### DISCUSSION TITLE
How to compute total delay in samples?

### DISCUSSION THREAD

**Discussion by pfeatherstone** [jgaeddert/liquid-dsp#270] [How do I....?] [answered]
Supposing we are chaining many filters, resamplers, equalizers, etc together, all of which have some delay.
How do you compute the total delay?
Supposing we have:
```
auto h1 = msresamp_crcf_create(...);
auto h2 = msresamp_crcf_create(...);
...
auto hN = msresamp_crcf_create(...);
```

How do you compute the delay of output samples when passing inputs into h1 -> h2 -> ... -> hN ?

**Comment by zacharyr** [comment]
If you have the original signal, you can compute the delay by using a correlation method against the delayed signal. `alignsignals` in MATLAB does this: https://www.mathworks.com/help/signal/ref/alignsignals.html. I have used this method successfully in the past for BER calculations and such.

**Comment by jgaeddert** [comment, marked as answer]
Let's assume that `h1` has a resampling rate `r1` and a delay `d1`. The output time index `t1` with respect to the original time index `t` is:

    t1 = (t + d1)*r1

So if you cascade resamplers back to back, the end delay would be

    tN = ((((t + d1)*r1 + d2)*r2 + d3)*r3 ... + dN)*rN
